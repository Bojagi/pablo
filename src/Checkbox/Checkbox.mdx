---
name: Checkbox
route: /checkbox
menu: Components
---

import { Props, Playground } from 'docz'
import { Box } from '../Box/Box.tsx'
import { Checkbox } from './Checkbox.tsx'

# Checkbox

Form element that allows users to select multiple items from a set of options. 
The checkbox shares an interface with the `Radio` and `Switch` components.

## Examples

### Single checkbox

<Playground>
  {() => {
    const [checked, setChecked] = React.useState(false);
    return (
      <Checkbox checked={checked} onChange={() => setChecked(!checked)} />
    );
  }}
</Playground>

### Checkboxes with labels

You can pass an optional `label` prop with a `ReactNode` value to show a label next to the checkbox.

<Playground>
  {() => {
    const [checked, setChecked] = React.useState(false);
    const [checkedTwo, setCheckedTwo] = React.useState(true);
    return (
      <Box display="flex" mx={-2}>
        <Checkbox mx={0.5} label="Spaghetti" checked={checked} onChange={() => setChecked(!checked)} />
        <Checkbox mx={0.5} label="Pizza" checked={checkedTwo} onChange={() => setCheckedTwo(!checkedTwo)} />
      </Box>
    );
  }}
</Playground>

### Small checkboxes

You can pass the `size` prop with the `small` value which renders a smaller version.

<Playground>
  {() => {
    const [checked, setChecked] = React.useState(false);
    const [checkedTwo, setCheckedTwo] = React.useState(true);
    return (
      <Box display="flex" mx={-2}>
        <Checkbox mx={0.5} size="small" checked={checked} onChange={() => setChecked(!checked)} />
        <Checkbox mx={0.5} size="small" checked={checkedTwo} onChange={() => setCheckedTwo(!checkedTwo)} />
      </Box>
    );
  }}
</Playground>

## Import

```js
import { Checkbox } from '@bojagi/pablo/Checkbox';
```

## Props
<Props of={Checkbox} />
