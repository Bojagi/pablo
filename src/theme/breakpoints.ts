export type Breakpoint = 'sm' | 'md' | 'lg' | 'xl';

export const breakpointNames: Breakpoint[] = ['sm', 'md', 'lg', 'xl'];

export type BreakpointsArray = [number, number, number, number];

export type Breakpoints = BreakpointsArray &
  Record<Breakpoint, number> & {
    breakpointNames: Breakpoint[];
  };

export function createBreakpoints(
  array: BreakpointsArray,
  usedBreakpointNames: Breakpoint[] = breakpointNames
): Breakpoints {
  const newBreakpoints = ([...array] as unknown) as Breakpoints;
  usedBreakpointNames.forEach((bp, index) => {
    newBreakpoints[bp] = newBreakpoints[index];
  });
  newBreakpoints.breakpointNames = usedBreakpointNames;

  return newBreakpoints as Breakpoints;
}

export const defaultBreakpoints: BreakpointsArray = [600, 1000, 1200, 1920];

export const breakpoints = createBreakpoints(defaultBreakpoints, breakpointNames);
