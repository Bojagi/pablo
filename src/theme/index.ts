export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
