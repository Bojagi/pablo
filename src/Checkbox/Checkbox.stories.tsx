import React, { useState } from 'react';
import { css } from '@emotion/react';
import { Flex, Box } from '../Box';
import { Checkbox, CheckboxProps } from './Checkbox';

export default {
  title: 'Checkbox',
};

const ControlledCheckbox = ({
  checked: checkedInitial,
  ...props
}: Omit<CheckboxProps, 'onChange'>) => {
  const [checked, setChecked] = useState(checkedInitial);
  return <Checkbox checked={checked} onChange={() => setChecked(!checked)} {...props} />;
};

const baseStory = (args) => (
  <Flex mx={-4} mb={4}>
    <ControlledCheckbox mx={4} checked={true} {...args} />
    <ControlledCheckbox mx={4} checked={false} {...args} />
    <ControlledCheckbox mx={4} checked={true} disabled={true} {...args} />
  </Flex>
);

export const Medium = baseStory.bind(null);
Medium.args = {
  size: 'medium',
};

export const Small = baseStory.bind(null);
Small.args = {
  size: 'small',
};

export const WithLabels = () => (
  <>
    <Box mb={4}>
      <ControlledCheckbox mb={4} checked={true} label="Hans" />
      <ControlledCheckbox mb={4} checked={false} label="Wurst" />
    </Box>
    <Box mb={4}>
      <ControlledCheckbox mb={4} checked={true} size="small" label="Hans" />
      <ControlledCheckbox mb={4} checked={false} size="small" label="Wurst" />
    </Box>
  </>
);

const customStyles = {
  box: css`
    background-color: blue;
  `,
  handle: css`
    background-color: red;
  `,
  label: css`
    text-decoration: underline;
  `,
};

export const WithCustomStyles = () => (
  <>
    <Box mb={4}>
      <ControlledCheckbox mb={4} checked={true} label="Hans" customStyles={customStyles} />
      <ControlledCheckbox mb={4} checked={false} label="Wurst" customStyles={customStyles} />
    </Box>
    <Box mb={4}>
      <ControlledCheckbox
        mb={4}
        checked={true}
        size="small"
        label="Hans"
        customStyles={customStyles}
      />
      <ControlledCheckbox
        mb={4}
        checked={false}
        size="small"
        label="Wurst"
        customStyles={customStyles}
      />
    </Box>
  </>
);
