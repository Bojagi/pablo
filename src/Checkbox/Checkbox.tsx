import React from 'react';
import styled, { css } from 'styled-components';
import { Box, BoxProps } from '../Box';
import { getComponentStyle, transitionTransformer } from '../utils/styleHelpers/getComponentStyle';
import { useComponentStyleContext } from '../theme';
import { Typography } from '../Typography';
import { useUniqueId } from '../utils/useUniqueId';
import { hijackCbBefore } from '../utils/hijackCb';

export type CheckboxSize = 'small' | 'medium';

interface CheckboxBoxProps {
  disabled?: boolean;
  size?: CheckboxSize;
  focus: boolean;
}

const CheckboxBox = styled.div<CheckboxBoxProps>`
  position: relative;
  width: ${getComponentStyle('checkbox.handleSize.{size}')};
  height: ${getComponentStyle('checkbox.handleSize.{size}')};
  border-radius: ${getComponentStyle('checkbox.outerBorderRadius')};
  padding: ${getComponentStyle('checkbox.innerPadding.{size}')};
  background-color: ${getComponentStyle('checkbox.backgroundColor')};
  border: ${getComponentStyle('checkbox.borderWidth')}px solid
    ${getComponentStyle('checkbox.borderColor')};
  opacity: ${(props) => (props.disabled ? 0.3 : 1)};

  transition: ${getComponentStyle('checkbox.boxTransition', transitionTransformer)};

  ${(props) =>
    props.focus &&
    css`
      box-shadow: 0 0 0 ${getComponentStyle('checkbox.focus.outlineSize')}
        ${getComponentStyle('checkbox.focus.outlineColor')};
    `};
  }
`;

const HiddenInput = styled.input.attrs({ type: 'checkbox' })`
  position: absolute;
  display: block;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
  ${(props) =>
    !props.disabled &&
    css`
      cursor: pointer;
    `}
`;

interface CheckboxHandleProps {
  disabled?: boolean;
  checked: boolean;
  size?: CheckboxSize;
}

const CheckboxHandle = styled.div<CheckboxHandleProps>`
  width: ${getComponentStyle('checkbox.handleSize.{size}')};
  height: ${getComponentStyle('checkbox.handleSize.{size}')};
  transform: scale(${(props: any) => (props.checked ? 1 : 0)});
  border-radius: ${getComponentStyle('checkbox.innerBorderRadius')};
  transition: ${getComponentStyle('checkbox.handleTransition')};
  background-color: ${getComponentStyle('checkbox.handleColor')};
`;

export interface CheckboxProps extends BoxProps {
  id?: string;
  className?: string;
  disabled?: boolean;
  name?: string;
  size?: CheckboxSize;
  label?: React.ReactNode;
  checked: boolean;
  onChange: (value: string, e: React.FormEvent<HTMLInputElement>) => void;
  onClick?: (e: React.FormEvent<HTMLInputElement>) => void;
  onFocus?: () => void;
  onBlur?: () => void;
}

export const Checkbox = ({
  id: idProp,
  className,
  size = 'medium',
  name,
  disabled,
  checked,
  label,
  onChange,
  onClick,
  onFocus,
  onBlur,
  ...props
}: CheckboxProps) => {
  const generatedId = useUniqueId('checkbox');
  const id = idProp || generatedId;
  const componentStyles = useComponentStyleContext();
  const typographyVariant = componentStyles.checkbox.typographyVariant[size];
  const [focus, setFocus] = React.useState(false);

  return (
    <Box display="flex" alignItems="center" {...props}>
      <CheckboxBox
        data-testid="pbl-checkbox"
        className={className}
        disabled={disabled}
        size={size}
        focus={focus}
        {...props}
      >
        <CheckboxHandle data-testid="pbl-checkbox-handle" checked={checked} size={size} />
        <HiddenInput
          id={id}
          data-testid="pbl-checkbox-input"
          name={name}
          checked={checked}
          disabled={disabled}
          onChange={
            onChange
              ? (e) => {
                  onChange(e.target.value, e);
                }
              : undefined
          }
          onClick={onClick}
          onFocus={hijackCbBefore(onFocus, () => setFocus(true))}
          onBlur={hijackCbBefore(onBlur, () => setFocus(false))}
        />
      </CheckboxBox>
      {label && (
        <label data-testid="pbl-checkbox-label" htmlFor={id}>
          <Typography ml={1} mb={0} variant={typographyVariant}>
            {label}
          </Typography>
        </label>
      )}
    </Box>
  );
};
