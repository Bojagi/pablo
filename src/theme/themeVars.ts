import isObject from 'isobject';
import { breakpointVars } from './breakpoints';
import { colorVars } from './colors';
import { typographyVars } from './typography';

export const themeVarNames = {
  typography: typographyVars,
  colors: colorVars,
  breakpoints: breakpointVars,
};

function createThemeReferences<T extends Record<string, any>>(obj: T): T {
  return Object.entries(obj).reduce(
    (acc, [key, value]) => ({
      ...acc,
      [key]: isObject(value) ? createThemeReferences(value as T) : `var(--${value})`,
    }),
    {} as any
  );
}

export const themeVars = createThemeReferences(themeVarNames);
