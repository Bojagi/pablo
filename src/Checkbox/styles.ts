import { Style } from '../theme/types';
import { checkableBaseStyles, CheckableBaseStyles } from '../shared/checkableBaseStyles';
import { themeVars } from '../theme/themeVars';

export interface CheckboxStyles extends CheckableBaseStyles {
  handleColor: Style;
  innerBorderRadius: Style | string;
  outerBorderRadius: Style | string;
}

export const checkboxStyles: CheckboxStyles = {
  ...checkableBaseStyles,
  handleColor: themeVars.colors.brand.main,
  innerBorderRadius: '3px',
  outerBorderRadius: '4px',
};
