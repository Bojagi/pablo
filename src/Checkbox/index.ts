import './styles';

export * from './Checkbox';
