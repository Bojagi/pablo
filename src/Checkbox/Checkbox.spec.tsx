import { act, cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';
import { PabloThemeProvider } from '../theme';
import { themeVars } from '../theme/themeVars';
import { Checkbox } from './Checkbox';

let onChangeMock;

beforeEach(() => {
  onChangeMock = jest.fn();
});

afterEach(cleanup);

test('Render checkbox unchecked', () => {
  const { container, getByTestId } = renderComponent({ checked: false });
  expect(container).toMatchSnapshot();
  expect(getByTestId('pbl-checkbox-handle')).toHaveStyleRule('transform', 'scale(0)');
  expect(getByTestId('pbl-checkbox-input')).not.toBeChecked();
});

test('Render checkbox checked', () => {
  const { container, getByTestId } = renderComponent({ checked: true });
  expect(container).toMatchSnapshot();
  expect(getByTestId('pbl-checkbox-handle')).toHaveStyleRule('transform', 'scale(1)');
  expect(getByTestId('pbl-checkbox-input')).toBeChecked();
});

test('Render checkbox with label', () => {
  const { getByTestId } = renderComponent({
    checked: true,
    label: (
      <>
        I am a <strong>label</strong>
      </>
    ),
  });
  expect(getByTestId('pbl-checkbox-label')).toBeDefined();

  // Label and input are connected via "id"
  expect(getByTestId('pbl-checkbox-label')).toHaveAttribute('for', 'custom-id');
  expect(getByTestId('pbl-checkbox-input')).toHaveAttribute('id', 'custom-id');
  expect(getByTestId('pbl-checkbox-label')).toMatchSnapshot();
  expect(getByTestId('pbl-checkbox-label')).toHaveTextContent('I am a label');
});

test('Click on checkbox to trigger onChange event', () => {
  const { getByTestId } = renderComponent({
    checked: false,
    value: 'checky',
    label: 'i am a label',
  });
  expect(onChangeMock).toHaveBeenCalledTimes(0);
  expect(getByTestId('pbl-checkbox-input')).toBeEnabled();
  act(() => {
    fireEvent.click(getByTestId('pbl-checkbox-input'), { target: { value: 'checky' } });
  });
  expect(onChangeMock).toHaveBeenCalledTimes(1);
  expect(onChangeMock).toHaveBeenCalledWith('checky', expect.objectContaining({ type: 'change' }));
});

test('Disabled prop should disable input', () => {
  const { getByTestId } = renderComponent({
    checked: true,
    disabled: true,
  });
  expect(getByTestId('pbl-checkbox-input')).toBeDisabled();
});

test('Render medium checkbox', () => {
  const { getByTestId } = renderComponent({
    size: 'medium',
    label: 'medium label',
  });
  expect(getByTestId('pbl-checkbox')).toHaveStyleRule(
    'width',
    /calc\(clamp\(0.75rem, .*?, 1.5rem\) \+ 2 \* \(clamp\(0.25rem, .*?, 0.5rem\) \+ 1px\)\)/
  );
  expect(getByTestId('pbl-checkbox')).toHaveStyleRule(
    'height',
    /calc\(clamp\(0.75rem, .*?, 1.5rem\) \+ 2 \* \(clamp\(0.25rem, .*?, 0.5rem\) \+ 1px\)\)/
  );
  expect(getByTestId('pbl-checkbox')).toHaveStyleRule('padding', /clamp\(0.25rem, .*?, 0.5rem\)/);
  expect(getByTestId('pbl-checkbox-handle')).toHaveStyleRule(
    'width',
    /clamp\(0.75rem, .*?, 1.5rem\)/
  );
  expect(getByTestId('pbl-checkbox-handle')).toHaveStyleRule(
    'height',
    /clamp\(0.75rem, .*?, 1.5rem\)/
  );
  // Uses subtitle typography
  expect(getByTestId('pbl-checkbox-label')).toHaveStyleRule(
    'font-size',
    themeVars.typography.body.fontSize
  );
});

test('Render small checkbox', () => {
  const { getByTestId } = renderComponent({
    size: 'small',
    label: 'small label',
  });
  expect(getByTestId('pbl-checkbox')).toHaveStyleRule(
    'width',
    /calc\(clamp\(0.5rem, .*?, 1rem\) \+ 2 \* \(clamp\(0.25rem, .*?, 0.5rem\) \+ 1px\)\)/
  );
  expect(getByTestId('pbl-checkbox')).toHaveStyleRule(
    'height',
    /calc\(clamp\(0.5rem, .*?, 1rem\) \+ 2 \* \(clamp\(0.25rem, .*?, 0.5rem\) \+ 1px\)\)/
  );
  expect(getByTestId('pbl-checkbox')).toHaveStyleRule('padding', /clamp\(0.25rem, .*?, 0.5rem\)/);
  expect(getByTestId('pbl-checkbox-handle')).toHaveStyleRule('width', /clamp\(0.5rem, .*?, 1rem\)/);
  expect(getByTestId('pbl-checkbox-handle')).toHaveStyleRule(
    'height',
    /clamp\(0.5rem, .*?, 1rem\)/
  );
  // Uses paragraph typography
  expect(getByTestId('pbl-checkbox-label')).toHaveStyleRule(
    'font-size',
    themeVars.typography.body.fontSize
  );
});

test('Use generated id when no "id" prop is set', () => {
  const { getByTestId } = renderComponent({
    id: undefined,
  });

  expect(getByTestId('pbl-checkbox-input')).toHaveAttribute('id', expect.stringMatching(/:.*?:/));
});

test('Render without onChange function ', () => {
  renderComponent({
    onChange: undefined,
  });
});

test('Forward onFocus and onBlur to input and show focus outline', () => {
  const onFocusMock = jest.fn();
  const onBlurMock = jest.fn();
  const { getByTestId } = renderComponent({
    onChange: undefined,
    checked: undefined,
    onFocus: onFocusMock,
    onBlur: onBlurMock,
  });

  expect(onFocusMock).toHaveBeenCalledTimes(0);
  act(() => {
    fireEvent.focus(getByTestId('pbl-checkbox-input'));
  });
  expect(onFocusMock).toHaveBeenCalledTimes(1);

  expect(getByTestId('pbl-checkbox')).toHaveStyleRule(
    'box-shadow',
    `0 0 0 3px ${themeVars.colors.brand.lightest}`
  );

  expect(onBlurMock).toHaveBeenCalledTimes(0);
  act(() => {
    fireEvent.blur(getByTestId('pbl-checkbox-input'));
  });
  expect(onBlurMock).toHaveBeenCalledTimes(1);

  expect(getByTestId('pbl-checkbox')).not.toHaveStyleRule(
    'box-shadow',
    `0 0 0 3px ${themeVars.colors.brand.lightest}`
  );
});

function renderComponent(props) {
  return render(
    <PabloThemeProvider>
      <Checkbox id="custom-id" onChange={onChangeMock} {...props}>
        Hello
      </Checkbox>
    </PabloThemeProvider>
  );
}
