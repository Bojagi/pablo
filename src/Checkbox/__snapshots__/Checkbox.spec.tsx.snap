// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render checkbox checked 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  position: absolute;
  display: block;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
  cursor: pointer;
}

.c1 {
  position: relative;
  width: 12px;
  height: 12px;
  border-radius: 4px;
  padding: 6px;
  background-color: #ffffff;
  border: 1px solid #00000040;
  opacity: 1;
  -webkit-transition: box-shadow 0.3s ease-in-out;
  transition: box-shadow 0.3s ease-in-out;
}

.c2 {
  width: 12px;
  height: 12px;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  border-radius: 3px;
  -webkit-transition: -webkit-transform0.15scubic-bezier(0.34,1.56,0.64,1);
  -webkit-transition: transform0.15scubic-bezier(0.34,1.56,0.64,1);
  transition: transform0.15scubic-bezier(0.34,1.56,0.64,1);
  background-color: #6A50F2;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="pbl-checkbox"
    >
      <div
        class="c2"
        data-testid="pbl-checkbox-handle"
      />
      <input
        checked=""
        class="c3"
        data-testid="pbl-checkbox-input"
        id="custom-id"
        type="checkbox"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Render checkbox unchecked 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  position: absolute;
  display: block;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
  cursor: pointer;
}

.c1 {
  position: relative;
  width: 12px;
  height: 12px;
  border-radius: 4px;
  padding: 6px;
  background-color: #ffffff;
  border: 1px solid #00000040;
  opacity: 1;
  -webkit-transition: box-shadow 0.3s ease-in-out;
  transition: box-shadow 0.3s ease-in-out;
}

.c2 {
  width: 12px;
  height: 12px;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  border-radius: 3px;
  -webkit-transition: -webkit-transform0.15scubic-bezier(0.34,1.56,0.64,1);
  -webkit-transition: transform0.15scubic-bezier(0.34,1.56,0.64,1);
  transition: transform0.15scubic-bezier(0.34,1.56,0.64,1);
  background-color: #6A50F2;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      data-testid="pbl-checkbox"
    >
      <div
        class="c2"
        data-testid="pbl-checkbox-handle"
      />
      <input
        class="c3"
        data-testid="pbl-checkbox-input"
        id="custom-id"
        type="checkbox"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Render checkbox with label 1`] = `
.c0 {
  font-family: "IBM Plex Sans",sans-serif;
  font-style: normal;
  font-weight: normal;
  margin: 0;
  font-size: 1rem;
  line-height: 1.375em;
  margin-bottom: 0.5em;
  margin-bottom: 0px;
  margin-left: 8px;
}

<label
  data-testid="pbl-checkbox-label"
  for="custom-id"
>
  <h4
    class="c0"
  >
    I am a 
    <strong>
      label
    </strong>
  </h4>
</label>
`;
